src/VisualStudio/TestUtilities/IntegrationHost.cs

```diff
@@ -0,0 +1,326 @@
+using System.Threading.Tasks;
+using System.Windows.Automation;
+using EnvDTE;
+using Microsoft.Win32;
+
+using DteProject = EnvDTE.Project;
+using Process = System.Diagnostics.Process;
+
+namespace Roslyn.VisualStudio.Test.Utilities
+{
+    public class IntegrationHost : IDisposable
+    {
+        internal static readonly string VsProductVersion = Settings.Default.VsProductVersion;
+        internal static readonly string VsProgId = $"VisualStudio.DTE.{VsProductVersion}";
+
+        internal static readonly string Wow6432Registry = Environment.Is64BitProcess ? "WOW6432Node" : string.Empty;
+        internal static readonly string VsRegistryRoot = Path.Combine("SOFTWARE", Wow6432Registry, "Microsoft", "VisualStudio", VsProductVersion);
+        internal static readonly string VsConfigRootSubKey = $"{VsRegistryRoot}_Config";
+        internal static readonly string VsInitConfigSubKey = Path.Combine(VsConfigRootSubKey, "Initialization");
+
+        internal static readonly string VsCommon7Folder = Path.GetFullPath(IntegrationHelper.GetRegistryKeyValue(Registry.LocalMachine, VsRegistryRoot, "InstallDir").ToString());
+        internal static readonly string VsUserFilesFolder = Path.GetFullPath(IntegrationHelper.GetRegistryKeyValue(Registry.CurrentUser, VsInitConfigSubKey, "UserFilesFolder").ToString());
+
+        internal static readonly string VsExeFile = Path.Combine(VsCommon7Folder, "devenv.exe");
+        internal static readonly string VsLaunchArgs = $"{(string.IsNullOrWhiteSpace(Settings.Default.VsRootSuffix) ? "/log" : $"/rootsuffix {Settings.Default.VsRootSuffix}")} /log";
+
+        internal static readonly string VsStartServiceCommand = "Tools.StartIntegrationTestService";
+        internal static readonly string VsStopServiceCommand = "Tools.StopIntegrationTestService";
+
+        private InteractiveWindow _csharpInteractiveWindow;
+        private DTE _dte;
+        private EditorWindow _editorWindow;
+        private Process _hostProcess;
+        private IntegrationService _service;
+        private IpcClientChannel _serviceChannel;
+        private SolutionExplorer _solutionExplorer;
+        private string _serviceUri;
+        private bool _requireNewInstance;
+
+        static IntegrationHost()
+        {
+            // Enable TraceListenerLogging here since we can have multiple instances of IntegrationHost created per process, but this is a one-time setup
+            Trace.Listeners.Clear();
+            Trace.Listeners.Add(new IntegrationTraceListener());
+        }
+
+        public IntegrationHost() { }
+
+        ~IntegrationHost()
+        {
+            Dispose(false);
+        }
+
+        public DTE Dte
+            => _dte;
+
+        public InteractiveWindow CSharpInteractiveWindow
+        {
+            get
+            {
+                if (_csharpInteractiveWindow == null)
+                {
+                    _csharpInteractiveWindow = InteractiveWindow.CreateCSharpInteractiveWindow(this);
```

Comment:
Rather than coupling these two classes together, perhaps InteractiveWindow.GetCSharpInteractiveWindow() could be the static method, and _that_ internally has a conditional weak table or a property bag or something? My worry is that as this expands, this class is going to grow all sorts of tentacles.